File: frontend/src/sim/ai/fusion.ts
import type { SensorSimulationResult } from "../sensors/base";
import { SENSOR_REGISTRY } from "../sensors/registry";

export interface FusionResult {
  fusedProbability: Float64Array;
  agreementCount: Int32Array;
  sensorKeys: string[];
  weights: Record<string, number>;
}

export function fuseSensorResults(results: Record<string, SensorSimulationResult>): FusionResult {
  const keys = Object.keys(results);
  if (keys.length === 0) throw new Error("At least one sensor result is required for fusion");

  const rawWeights = keys.map((k) => SENSOR_REGISTRY[k].base_reliability);
  const weightSum = rawWeights.reduce((a, b) => a + b, 0);
  const weights = rawWeights.map((w) => w / weightSum);

  const n = results[keys[0]].probabilityVolume.length;
  const fused = new Float64Array(n);
  const agreement = new Int32Array(n);

  for (let k = 0; k < keys.length; k++) {
    const vol = results[keys[k]].probabilityVolume;
    const w = weights[k];
    for (let i = 0; i < n; i++) {
      fused[i] += vol[i] * w;
      if (vol[i] > 0.5) agreement[i] += 1;
    }
  }

  const weightMap: Record<string, number> = {};
  keys.forEach((k, i) => (weightMap[k] = weights[i]));

  return { fusedProbability: fused, agreementCount: agreement, sensorKeys: keys, weights: weightMap };
}

File: backend/app/ai/fusion.py
"""Multi-sensor data fusion.

Combines every active sensor's per-voxel anomaly-probability volume into a
single fused probability volume, weighting each sensor by its typical
reliability. Also tracks a per-voxel "agreement count" — how many sensors
independently flagged that voxel — which is a strong, cheap confidence
signal in real multi-sensor geophysical surveys.
"""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from app.sensors.base import SensorSimulationResult
from app.sensors.registry import SENSOR_REGISTRY


@dataclass
class FusionResult:
    fused_probability: np.ndarray
    agreement_count: np.ndarray
    sensor_keys: list[str]
    weights: dict[str, float]


def fuse_sensor_results(results: dict[str, SensorSimulationResult]) -> FusionResult:
    if not results:
        raise ValueError("At least one sensor result is required for fusion")

    keys = list(results.keys())
    volumes = np.stack([results[k].probability_volume for k in keys], axis=0)
    raw_weights = np.array([SENSOR_REGISTRY[k].base_reliability for k in keys])
    weights = raw_weights / raw_weights.sum()

    fused = np.tensordot(weights, volumes, axes=([0], [0]))
    agreement = (volumes > 0.5).sum(axis=0)

    return FusionResult(
        fused_probability=fused,
        agreement_count=agreement,
        sensor_keys=keys,
        weights=dict(zip(keys, weights.tolist())),
    )
